### ISSUE TITLE
Unable to Authorize Works with Nest via NST Manager

### ISSUE THREAD

**Issue by rbiffle** [tonesto7/nest-manager#442]
## Verify the following before opening an issue:

- [X] That _OAuth_ is Enabled for the SmartApp under the IDE.
- [X] The SmartApp and Device are using the latest version.
- [X] You have reviewed the Readme for potential fixes.
- [X] That the Device Types have been _Published for You_.

## About Your Setup:
* How many Thermostats?: 3
* How many Protects? 0
* Are you Using the Nest Presence Device? No
* iOS or Android?: iOS
* Mobile App Version:5.0.3

## Expected Behavior
 Expecting the application to open the Works with Nest Window, allow me to sign in to my Nest account, then allow me to authorize the ST and Nest connection.  

## Current Behavior
Each time the NST Manager asks me to sign into my Nest account and asks me to allow the Nest and ST connection, there is an error at the top of the Works with Nest page.  It says one of two things, Oops we experienced an error, try again later, or We cannot reach Nest.com, please try again later.  My application was sending numerous Warning Offline message so I thought I would remove all NST apps and reinstall.  I do not have any automations so it was not supposed to be a big deal.  When I could not get the Works with Nest page to allow the ST connection, I tried to uninstall again and perform the reinstall.  Now I cannot get the NST Automations app to delete via IDE because it says one or more users have it installed.  But it will not allow me to uninstall or remove from he the ST Application.  So there are two points to my issue: 1.  Can you tell me how to delete the NST Automations from Smartthings?  2.) What could be causing Nest to not allow the ST connection?

## Steps to Reproduce (for bugs)

1.
2.
3.
4.

## Context
  Clean uninstall and clean reinstall of the NST Manager and NST Automations.

Please include a copy of your log under the IDE if possible>>>

**Comment by rbiffle** [comment]
Ok, so update to this issue.  Successfully uninstalled the the NST Manager and NST Automations using instructions that @tonesto7 sent to another user.  I reinstalled per the instructions, OAuth is enabled, and all smart apps and device handlers are reinstalled.  I've published for me and saved.  I only installed the Nest Thermostat (I have 3), Weather, and Presence.  I've checked my Nest account and see no Works with Nest connections.  ST is not listed.  Is there something I am missing with the Works with Nest?  I still get the "Oops! We encountered an error. Please try again" error message.  Is there something that you @tonesto7 need to reset to allow me to authorize the Works with Nest connection?

**Comment by wrightinthe6ix** [comment]
I am having the same issue.. I have installed and I can not click the Accept button in the Login to next section of the Nest Manager App. 
I have uninstalled and reinstalled as above.. and ensured the the OAuth is enabled. 
Any suggestions?

**Comment by bhupesh102** [comment]
Having Similar issue,

Cant get nest to authenticate NST manager. After I login, clicking the Blue Accept button on Works with nest page does nothing. There is an error on the top "Oops! We encountered an error. Please try again later. I have tried multiple uninstalls and installs and have oAuth enabled. Please help.

Log is attached.
![image](https://cloud.githubusercontent.com/assets/27539360/25189516/fc250de8-24ee-11e7-880b-e9e651328f22.png)

Thanks,

Bhupesh

**Comment by tonesto7** [comment]
We ran out of auth tokens.  We are temporarily remeding the situation.

**Comment by bhupesh102** [comment]
Thank you for letting us know. How long should this take? We appreciate all your efforts.

Bhupesh

**Comment by rko180** [comment]
Hi @tonesto7 I'm having this same issue. Does it has to do with the tokens?

**Comment by stubn** [comment]
I'm having this issue too, is there a resolution?

**Comment by Tokn59** [comment]
Hello, I also have this issue. Cannot login in the NST Manager "Login to Nest" - error : Oops ... we've encountered ....

OAuth is Enabled for the SmartApp under the IDE. - **Checked**
The SmartApp and Device are using the latest version. - **Updated from repo just now**
You have reviewed the Readme for potential fixes. - **Yes**
That the Device Types have been Published for You. - **Yes they are (see screenshot)**
![nest issue 1](https://cloud.githubusercontent.com/assets/9215231/26830014/bab9f78e-4ac7-11e7-9de4-6a6aaa299fae.JPG)
![nest issue 2](https://cloud.githubusercontent.com/assets/9215231/26830013/bab95a90-4ac7-11e7-9066-87f9c803a3a4.JPG)

**
How many Thermostats?: **0**
How many Protects? **2**
Are you Using the Nest Presence Device? **No**
iOS or Android?: **iOS**
Mobile App Version:**5.0.9**

Thank you for your advice and help ! **Will await remedy with tokens !**

Kind Regards, Tony

**Comment by uberrob** [comment]
Hi @tonesto7 please reopen this ticket,  I am having the same issue. 

Cannot login in the NST Manager "Login to Nest" - error : Oops ... we've encountered ....

OAuth is Enabled for the SmartApp under the IDE. - Checked
The SmartApp and Device are using the latest version. - Updated from repo just now
You have reviewed the Readme for potential fixes. - Yes
That the Device Types have been Published for You. - Yes they are

Thank yiu

Rob

**Comment by jwhazzard** [comment]
Hi @tonesto7, I am also receiving the error "Oops! We encountered an error. Please try again." when clicking "Accept" after the Nest login. I have deleted and added back the app and receive the same error. Any suggestions? Thanks!

65a5c543-d0aa-4587-afeb-2c0d0f0aa04b  2:38:54 PM: warn || Nest Manager | Missing: structData null
65a5c543-d0aa-4587-afeb-2c0d0f0aa04b  2:38:52 PM: info ||| Nest Manager | AuthToken not found: Directing to Login Page
65a5c543-d0aa-4587-afeb-2c0d0f0aa04b  2:38:52 PM: info ||| Nest Manager | appData.json File HAS Changed

**Comment by neosfranger** [comment]
I was having a similar issue, but now I am able to hit the "accept" button but then the app provides me with a Nest Pin code, yet I am never prompted to enter the pin to authorize the account. What am I suppose to do with the Nest Pincode or should I be inputing something in the OAuth Redirect URI section?

**Comment by imnotbob** [comment]
We have just updated the tokens, let us know if it is working now.

**Comment by jwhazzard** [comment]
I'm​ receiving the error below after clicking accept. Thanks for responding to the issue. 

Webpage not available

The webpage at https://graph.api.smarthings.com/oauth/callback?state=8bafed2b-b7b8-419f-b8a3-ec43bc1dbde1&code=FQEHXXWLNRVJ7K47 could not be loaded because:

net::ERR_CONNECTION_TIMED_OUT

**Comment by xomikronx** [comment]
Still getting "Oops! We encountered an error.  Please try again."

**Comment by hunt2714** [comment]
When trying to click the blue Accept button on the Works with Nest page, it says "Can't Reach nest.com right now. Try again soon." Is this relates to the lack of tokens or indicative of another issue?

**Comment by janselen** [comment]
I'm getting the same problem. "Oops! We encountered an error. Please try again."

**Comment by imnotbob** [comment]
should be fixed now

**Comment by antonvlad** [comment]
I just tried setting this up and I'm getting pretty much the same issue. There's nothing in the log beyond "AuthToken not found: Directing to Login Page"
When I log in  to Works with Nest and hit accept, I get "Can't reach nest.com right now. Try again soon."

**Comment by TommyFix** [comment]
I am also getting "Oops! We encountered an error. Please try again." I suppose we ran out of tokens?

**Comment by TommyFix** [comment]
Was able to authorize it just now! Cheers

**Comment by melissamriveraicloud** [comment]
I am having trouble authorizing. Any idea how often tokens are updated? Thank you for all the hard work!

**Comment by marshalk86** [comment]
I am having the same issue. I get this error message: "Oops! We encountered an error. Please try again."

Any updates on why this is happening?

**Comment by Misaka-12450** [comment]
Same issue here.

**Comment by ddebellis** [comment]
"Oops! We encountered an error. Please try again."

4:31:58 AM: warn || Nest Manager | Missing: structData null

**Comment by Juberstine** [comment]
Same issue for me

**Comment by imnotbob** [comment]
try now

**Comment by shumws** [comment]
Just reinstall the 5.1.8, having the same issue. Are we out of tokens?

16ce6b63-8f26-464c-9c72-4339eca3fb6f  3:15:07 PM: warn || Nest Manager | Missing: structData null
16ce6b63-8f26-464c-9c72-4339eca3fb6f  3:15:06 PM: info ||| Nest Manager | AuthToken not found: Directing to Login Page

**Comment by shumws** [comment]
Seems to be working now. Thanks!

**Comment by mp46241** [comment]
I am getting this issue, same error as others here, just installed this and haven't been able to get passed the works with nest page

**Comment by paisita1012** [comment]
Same issues as poster above. Thank you

**Comment by imnotbob** [comment]
check now

**Comment by paisita1012** [comment]
indeed. thank you!

**Comment by fstephanie** [comment]
I have the same issue as I try to hit "accept".  Can someone here help? Thank you

**Comment by cnetsols** [comment]
Hi,
I'm getting the message "Oops! We encountered an error. Please try again." after clicking on the Accept button. Any help? thanks!!!

**Comment by imnotbob** [comment]
try now

**Comment by cnetsols** [comment]
Thank you very much!!! It's working now.

**Comment by carrahar** [comment]
I have got this same issue now, pressing accept gives me the Opps error

thanks

**Comment by tonesto7** [comment]
@carrahar Please try again

**Comment by carrahar** [comment]
Hi still no joy i just tried a full uninstall and re install

**Comment by tonesto7** [comment]
Sorry about that.  The changes never saved to the DB.  It should work now

**Comment by carrahar** [comment]
superb its working now thank you very much for your support i appreciate your help

**Comment by gmcmenimen** [comment]
Getting the error again.  Clicking the Accept button gives the Oops error.

**Comment by imnotbob** [comment]
Please be sure you are running the latest version of code.

**Comment by Ysater** [comment]
Hi, 
First thing, thanks for providing a great solution and I appreciate all the hard work this must have been!

I am having trouble setting up the NST Manager, unfortunately due to my location i am unable to use the GITHUB interface to load it into my smart apps and had to use the manual method.

I have added both the NST Manager and NST Automations and enabled OAuth only to NST Manager.
After going through the setup on my ST App, i was able to use my Nest credentials and was able to see my 4 protects in NST Manager. 

However i later get a message: Device Handlers are Missing or Not Published. Please verify all device handlers are present before continuing. 

Also I do not see my devices on the smartthings devices list.

I am not sure what i did wrong, can you please help!

Thanks for your help!

**Comment by imnotbob** [comment]
See:

http://thingsthataresmart.wiki/index.php?title=NST_Manager

specifically:
http://thingsthataresmart.wiki/index.php?title=NST_Manager#Nest_Device_Installation

**Comment by IBMJim** [comment]
Getting the Oops error.

**Comment by ghost** [comment]
Same as everyone else here.  I am getting the "Oops! We encountered an error"

**Comment by 62165** [comment]
I'm also getting the "Oops!" error...I've tried to uninstall/reinstall...

**Comment by mslanh01** [comment]
Getting the same error as users above..

**Comment by imnotbob** [comment]
try now

**Comment by ghost** [comment]
Works now. Thanks.

**Comment by mslanh01** [comment]
It works! Thanks!

**Comment by IBMJim** [comment]
Thank you!!!!!

**Comment by nickfoti** [comment]
having this same issue as those above

**Comment by stevencanale** [comment]
having issue as well

**Comment by rdprosser** [comment]
ditto

**Comment by imnotbob** [comment]
try now

**Comment by rdprosser** [comment]
working thanks!

**Comment by nickfoti** [comment]
working now, thank you!

**Comment by lokak** [comment]
having same issue as the users above

**Comment by imnotbob** [comment]
try now

**Comment by dansingha** [comment]
We re getting this error "Oops  we encountered an error ...."
when we click on "Accept" button.

**Comment by daoom** [comment]
I am getting the "Oops! ... " error when clicking the accept button. the following is showing up in the live logging:

a1a77fe9-f158-4e7e-8624-fcd77e2fe6eb  02:12:55: warn || Nest Manager | Missing: structData null
a1a77fe9-f158-4e7e-8624-fcd77e2fe6eb  02:12:53: info ||| Nest Manager | AuthToken not found: Directing to Login Page
a1a77fe9-f158-4e7e-8624-fcd77e2fe6eb  02:12:53: info ||| Nest Manager | appData.json File HAS Changed

I also checked the following: 

- OAuth is Enabled for the SmartApp under the IDE : yes
- he SmartApp and Device are using the latest version: v5.2.0 (8-12-2017)
- You have reviewed the Readme for potential fixes: yes
- That the Device Types have been Published for You.: yes

About Your Setup:

- How many Thermostats?: 1
- How many Protects? 1
- Are you Using the Nest Presence Device? No
- iOS or Android?: Android
- Mobile App Version: 5.2.0

**Comment by imnotbob** [comment]
try now

**Comment by dodforensics** [comment]
I'm getting the same error as everyone else "Oops! We encountered an error. Please try again. when trying to accept.

**Comment by luxore** [comment]
Same here, let me know what I can provide to help troubleshoot

**Comment by daoom** [comment]
@imnotbob: I Tred it again, and it worked. 

Not sure what you did on your end, but thanks :)

**Comment by dodforensics** [comment]
Sadly still not working on my end

**Comment by luxore** [comment]
@tonesto7 I added 2 factor authentication to my nest account, closed the android app, reopened and tried adding the nest manager app. It seems that the addition of the 2FA allowed me to complete the connection process!

Thanks!

**Comment by dodforensics** [comment]
Nice! I enabled the 2FA and was able to successfully accept! All well in the world now

**Comment by clmcgrath** [comment]
Tried to do 2fa still can't accept ,  just gives me an error saying that an unknown error occured 
also i can't remove the app to try a reinstall

**Comment by sptjr1** [comment]
I to am running into the same issue.  I also have 2FA enabled.

**Comment by ibevamp** [comment]
Same issue here. I enabled 2FA and i am still unable to click accept. I started from scratch again, removed all repos, added then again, enabled OAuth, made sure everything was published. But i still cannot accept.

**Comment by frazzlegod** [comment]
Hi @tonesto7 and @imnotbob!

I've just followed all the instructions to install NST Manager (latest version 5.2.0 from repo) into SmartThings and I, too, get as far as the "Works with Nest" page, click on the "Accept" button, and just receive the infamous "Oops! We encountered an error" message.

So I can never authenticate with Nest using the NST Manager SmartApp and am at a dead end now it seems.

Please can you help me? I just bought a Nest Protect today (in UK btw), configured it using the Nest app, and now wanted to integrate with my SmartThings setup. I'm really keen to start using your NST Manager SmartApp as, from what I've seen and read so far, it's simply awesome.

I would appreciate any help you can provide to me.

I also just get the following in the SmartThings IDE Live Logging console:
21530cd9-0aa8-4844-b4ee-e6d58a317c39  7:30:44 PM: info ||| Nest Manager | AuthToken not found: Directing to Login Page


Many thanks in advance,
frazzlegod

**Comment by tonesto7** [comment]
Yeah sorry we are juggling tokens because of the 6200+ users. Nest will not approve us for more unless we lobotomized the app and devices.
I’ll be releasing instructions soon for end users to create there own tokens

Please give it a try now.

**Comment by sptjr1** [comment]
It worked for me.  Thank you so much for your help!

**Comment by frazzlegod** [comment]
Hi @tonesto7,

Thank you so much. I tried it again and it authenticated fine this time - my SmartThings and Nest are now successfully linked :) Thank you also for your really quick response, it is much appreciated.

I love the NST Manager SmartApp. So many configuration options, it's great. Really good work on the SmartApp!

Just a few questions about the token... Does my token have an expiry date/time? Will it suddenly expire one day? If that's the case, might I face the same authentication issue again in the NST Manager SmartApp in the future (trying with a fresh, new token)? Could this token auth issue manifest itself repeatedly, albeit on a hopefully infrequent basis?

Also, just a few observations from my really quick usage of the NST Manager and Nest device(s) in my SmartThings (I only have 1 Nest Protect at the moment, so can only comment on that 1 device type so far):

- When loading the Nest Protect device page (device type version 5.1.4), it loads already scrolled down the page, so the first main largest tile ("CLEAR") is already scrolled passed and off the page. This is with Android 7.0 .

-  Power Type is reported as "Null". My Nest Protect is on battery power.

- Tapping the "Refresh" tile appears to do nothing for a fair few seconds, then does refresh the page, for a split second shows the "Device currently unavailable" splash page, refreshes again and then displays hardly anything on the device's page, aside from the main largest tile ("CLEAR") and the "Refresh" tile, with a huge gap of white space in between (where the previous status information once was).

- Tapping the smaller blue "Refresh" button below the status information immediately results in a 404 server-y type page ("nginx"?) being displayed (completely replacing the status information boxes) every time.

- Navigating away from the "Right Now" view (to either "Recently" or "SmartApps") and back to the "Right Now" view makes the status information boxes disappear, leaving only the main largest tile ("CLEAR") and the "Refresh" tile, again with the huge gap of white space in between.


I don't know if such issues have already been raised elsewhere, or if you are aware of them, but just some of my observations upon using NST Manager and the Nest Protect device type for the first time in SmartThings. I hope you don't mind! :/

Anyway, thank you once again for sorting out my token auth issue, and it really is a great SmartApp 👍 👍 👍 

Many thanks,
frazzlegod

**Comment by battlebudy** [comment]
@tonesto7 I am getting the same thing as everyone else but i am running on android.

**Comment by scanzip** [comment]
I'm receiving the same problem.  Logging says "AuthToken not found" and I get the Ooops notification.

**Comment by clmcgrath** [comment]
After reinstalling everything and enabling  2fa i was able to finally get this working

**Comment by jikdoc** [comment]
Hi,
I'm having the issue with tokens.  is there a link to show how i can create my own?

**Comment by tonesto7** [comment]
There is a rough document outlining how to create one on the wiki.
http://thingsthataresmart.wiki/index.php?title=NST_Manager#Nest_Developer_Account

**Comment by raghavkashi** [comment]
I am getting the "Oops! We encountered an error. Please try again" message.. what should I be doing? Wait for more tokens to be available?

**Comment by imnotbob** [comment]
Please see the instructions:

http://thingsthataresmart.wiki/index.php?title=NST_Manager

specifically:

http://thingsthataresmart.wiki/index.php?title=NST_Manager#Installation_Instructions
note the section on Nest Developer Account

https://community.smartthings.com/t/release-nst-manager-v5-0/83228/733

The issue is Nest will not give us more tokens (you can get them yourself), because they view the tokens as not kept private enough. Smarthings will not assist us in keeping them private enough to meet Nest’s ask. So we are stuck between rock and hard place.

**Comment by raghavkashi** [comment]
I created a new product secret, but the OAuth is already populated on the NST smartapp and I am not able to edit it, nor I am able to delete the NST Manager from Smartapps in the ST IDE.

What are my options to update the product ID and secret now?

**Comment by raghavkashi** [comment]
Even after creating my own token and copying the product id and product secret to the smartapps settings as per the document, I still get the "Oops!! We've encoutered .." message when I click on "Accept"

**Comment by djtwreck** [comment]
Yep,,,, I went through the whole process of getting my token,,,, still getting the oops

**Comment by raghavkashi** [comment]
Tried again in the morning, and noticed the the product ID / secret was missing on the smartapp. Updated it and tried again and it worked fine. Now I can control my NEST thermostat from the ST app. Thanks for a great integration app

**Comment by djtwreck** [comment]
I also tried again this morning,,,,, uninstalled and reinstalled the smart app ,,,, made sure to update the client ID and secret I made from the nest developer,,,,, was able to sign in and everything works now,,,, thanks for a the great app!!!!

**Comment by prashanthrkallu** [comment]
@tonesto7, Thank you for the app and all your hard work. 

I have followed the wiki instructions to install my 
http://thingsthataresmart.wiki/index.php?title=NST_Manager#Installation_Instructions
Followed  below link to Create my ID and Token for Nest
https://docs.google.com/document/d/1JoqNijFvLAHaQIMrqHCSBXgglUqdEnj3rBD9EiPP3bs/edit

I have copied the product id and product secret to the smartapps settings as per the document, I still get the "Oops!! We've encoutered .." message when I click on "Accept".

It would be great if you can help me with this, I am new to this and am not sure if I am missing anything.

I have Verify the following before opening this issue:
 That OAuth is Enabled for the SmartApp under the IDE.
 The SmartApp and Device are using the latest version.
 You have reviewed the Readme for potential fixes.
 That the Device Types have been Published for You.

About My Setup:
How many Thermostats?: 1
How many Protects? 1
Are you Using the Nest Presence Device? No
iOS or Android?: iOS
Mobile App Version:5.0.3

**Comment by tonesto7** [comment]
Let me check the code but if ID and Secret are in the settings they are supposed to be used instead of built in

**Comment by laluprasad** [comment]
@tonesto7 , I followed same procedure at @prashanthrkallu   and I experience same error.  I don't think it looks like configuration issue but any help will be appreciated 

 8:50:49 AM: warn || Nest Manager | Missing: structData null
7fa6a8da-195f-45b7-acad-7360312c81e7  8:50:40 AM: info ||| Nest Manager | AuthToken not found: Directing to Login Page
7fa6a8da-195f-45b7-acad-7360312c81e7  8:50:01 AM: warn || Nest Manager | Missing: structData null
7fa6a8da-195f-45b7-acad-7360312c81e7  8:49:56 AM: info ||| Nest Manager | AuthToken not found: Directing to Login Page

**Comment by laluprasad** [comment]
@tonesto7 @prashanthrkallu   worked fine after I used Default OAuth Redirect URI
https://graph.api.smartthings.com/oauth/callback   instead of post march 2016 url .  Thanks

**Comment by prashanthrkallu** [comment]
@laluprasad,
Thank you. https://graph.api.smartthings.com/oauth/callback was not working for me while adding product but I went back and added this, it started working.

**Comment by jimy17j** [comment]
Please advise, i am getting this Oops message

**Comment by tinker2455** [comment]
I just installed NST Manager for the first time and am trying to integrate my Nest Protest into Smartthings.  When I try to Login to Nest I get the 'Oops! We encountered an error. Please try again.' message.  I followed installation instructions for NST Manager to the letter.  Please help!

**Comment by Coop901** [comment]
Having the same issue with mine.

**Comment by mferris2421** [comment]
I am also having the same issue; and when I tried to follow the instructions for my own app, it says the app already exists. The instructions appear to say to set the same name; so maybe they changed it or I should create my own.

**Comment by SSlimC** [comment]
Same problem guys, "Oops!We encountered an error. Please try again."
Thanks for any help you can offer

How many Thermostats?: 1
How many Protects? : 0
Are you Using the Nest Presence Device? Yes
iOS or Android?: Android
Mobile App Version:2.6.1

**Comment by Deja-Github** [comment]
Same issue here too.

**Comment by portauto** [comment]
Hello, I'm encountering the same issue, it says "Oops! We encountered an error. Please try again later" when I try to authenticate to Nest from the NST Manager.

That OAuth is Enabled for the SmartApp under the IDE. Yes
The SmartApp and Device are using the latest version. Yes
You have reviewed the Readme for potential fixes. Yes
That the Device Types have been Published for You. Yes

How many Thermostats?: 0
How many Protects? 7
Are you Using the Nest Presence Device? Yes
iOS or Android?: iOS
Mobile App Version:5.2.0

Live Logging:
9:18:10 PM: info ||| Nest Manager | AuthToken not found: Directing to Login Page
9:18:02 PM: info ||| Nest Manager | AuthToken not found: Directing to Login Page
9:18:01 PM: info ||| Nest Manager | appData.json File HAS Changed

**Comment by blazchasr** [comment]
Having the same issue even after following directions for creating my own clientid and clientsecret. Any help is appreciated

**Comment by mvpdms** [comment]
Just installed today but get the oops error when trying to accept the agreement through the nst manager app

**Comment by aplus2k** [comment]
I am also getting the Oops! We encountered an error message.

**Comment by mvpdms** [comment]
Once I created a nest developer account it worked. I originally had trouble
with the redirect url but after a few tries it worked.

**Comment by aplus2k** [comment]
I just tried creating a nest developer account and then tried again a few times.  I am still getting the Oops error message.

**Comment by bschnell90** [comment]
Hey @tonesto7 having the oops we encountered an error issue please help.

**Comment by nonvtec** [comment]
Same here @tonesto7, getting the oops we encountered an error.

**Comment by samywee** [comment]
Same here @tonesto7, getting the oops we encountered an error. (OAuth - enabled)

10:29:22 AM: info ||| Nest Manager | AuthToken not found: Directing to Login Page

**Comment by sebasrivard** [comment]
Looks promising but can't authenticate 
Same error as other 
Oops! We encountered an error. Please try again.

**Comment by samywee** [comment]
http://thingsthataresmart.wiki/index.php?title=NST_Manager#Installation_Instructions

May have to create a Nest Developer Account. Follow instructions..

**Comment by nonvtec** [comment]
Thanks samywee, I created the Nest Dev account and I’m good now, just got it working!

**Comment by sebasrivard** [comment]
That dir it for me too, fairly easy!
Should be in the guide as a mandatory step.
tsk for the goat work

**Comment by amchaud** [comment]
Good morning...
Get the same error message - Oops! We encountered an error. Please try again.
Latest versions 

That OAuth is Enabled for the SmartApp under the IDE. Yes
The SmartApp and Device are using the latest version. Yes
You have reviewed the Readme for potential fixes. Yes
That the Device Types have been Published for You. Yes

log:
7:54:24 AM: debug getChildDevices(false), children=2
7:54:24 AM: debug getChildDevices(false), children=2
7:54:24 AM: debug detecting status for 92fc25c3-6ae4-4daf-8800-99638f9a1522 : false
7:54:24 AM: trace in ssdpDiscover
7:53:49 AM: info ||| Nest Manager | AuthToken not found: Directing to Login Page
7:53:48 AM: info ||| Nest Manager | AuthToken not found: Directing to Login Page
7:53:48 AM: info ||| Nest Manager | AuthToken not found: Directing to Login Page
7:53:27 AM: debug Not parsing XML message:

I have a nest account developper...

can anybody help me ?

**Comment by bocam3** [comment]
Get the same error message - Oops! We encountered an error. Please try again.

Summary of progress:
That OAuth is Enabled for the SmartApp under the IDE. Yes
The SmartApp and Device are using the latest version. Yes
You have reviewed the Readme for potential fixes. Yes
That the Device Types have been Published for You. Yes
Nest Developer Account. Yes
NDA Password and Product Secret. Yes
Pasted Password/Prod Secret in Nest Manager App settings: Yes
Nest Manager Displayed in My Apps: Yes
Login to Nest with User Name and Password input: Yes
2FA text verification code received/entered: Yes
Nest Manager accepts verification code: No

I still get the "Oops! We encountered an error. Please try again."

UPDATE!
SEE PeterZiel tip below - make sure when you paste the NDA Password and Product Secret to remove an extra space at the end of each.

**Comment by PeterZiel** [comment]
I'm also having the same issue- Oops! We encountered an error. Please try again.

my progress is same as above
 That OAuth is Enabled for the SmartApp under the IDE. Yes
 The SmartApp and Device are using the latest version. Yes
 You have reviewed the Readme for potential fixes. Yes
 That the Device Types have been Published for You. Yes
 Nest Developer Account. Yes
 NDA Password and Product Secret. Yes
 Pasted Password/Prod Secret in Nest Manager App settings: Yes
 Nest Manager Displayed in My Apps: Yes
 Login to Nest with User Name and Password input: Yes
 2FA text verification code received/entered: Yes
 Nest Manager accepts verification code: No
I then tried uninstalling and reinstalling everything with the same result

is there anything I've missed and can remedy?

**Comment by PeterZiel** [comment]
I have found the solution:
When Pasting the Client ID & Secret into the NST Manager IDE make sure there isn't a space at the end of either code.  I simply removed the space, updated and it's working in my App.

hope this helps others

**Comment by bocam3** [comment]
@PeterZiel - BOOOM!  That was it.  I wonder if it makes sense to edit the Google doc with this tip.

**Comment by tinker2455** [comment]
An updated Google doc would help a bunch of us out!

**Comment by tonesto7** [comment]
I will add that note to the doc right now?

**Comment by tinker2455** [comment]
Excellent!  THANKS.  A link to the doc would also save some folks time searching for it.

**Comment by amchaud** [comment]
Hi

Finaly ibsilved my issue by creating a developper account with nest and using the client id and  the other number id within smarthings app that you developped.

Many thanks

Andre

**Comment by Hoveys** [comment]
@tonesto7 , can you reshare the doc.  it looks like Google removed it for terms of service violation.  Anyway to get it posted somwhere else?

![image](https://user-images.githubusercontent.com/15082192/32238033-8e1f0d50-be34-11e7-9f99-7732a8652557.png)

**Comment by dprofico1** [comment]
@tonesto7 I am getting the same issue when trying to log into Nest from the manager. oAuth is enabled I have read through the comments here and check to make sure I have set everything up. Could this be a token refresh issue?

**Comment by kencol78** [comment]
The 'Nest' shows up on the ST android app as 'Tiles Missing' and the NST Automations App shows 'Something's Wrong' !

Anyone know where I'm going wrong?

**Comment by myrison** [comment]
@tonesto7 - I've just installed for the first time as well today and have read through all of the open issues in this project and comments in this thread to make sure I'm not missing anything obvious.  I've set up my own Nest developer account and am using that info in the app, I've looked for trailing spaces in the configuration settings, checked that OAuth is enabled, etc., and still get the "oops something went wrong" when I try to authenticate in the Nest app.

I just included related comments in another issue here that is notifying of the same thing, but given this one appears far more active, I'll copy those comments over here if they're helpful:

> This appears to have something to do with the oauth redirect, but I'm admittedly guessing. I have the same problem as a few other posts here in "issues list."
> 
> The oauth 500 errors are described here:
> https://community.smartthings.com/t/status-code-500-from-oauth-token/58054
> 
> Whenever I try to access the oauth callback URL, I get a 500 error, which I'm sure is the cause of not being able to authenticate with Nest and get a token back. However, I can't figure out how to fix it.
> 
> It's the same error this user is having in another issue:
> #493
> 
> Whether you use this callback:
> https://graph-na04-useast2.api.smartthings.com/oauth/callback (my shard)
> 
> Or the generic one:
> https://graph.api.smartthings.com/oauth/callback

**Comment by kahnman** [comment]
Hi @tonesto7,

I am getting the same "Oops! We've encountered an error. Please try again." error that many people here are seeing when clicking the "Accept" button. I installed first the first time today using [these instructions](http://thingsthataresmart.wiki/index.php?title=NST_Manager#Installation_Instructions). I have tried the following:

- I've enabled OAuth for the "Nest Manager" SmartApp and tried both https://graph.api.smartthings.com/oauth/callback and https://graph-na02-useast1.api.smartthings.com/oauth/callback as the Redirect URI
- Using my own nest developer account, I've generated and place the clientId and clientSecret in the IDE. 
- I've enabled 2FA on my nest account. 

Once interesting thing that I've seen that I haven't seen anyone else mention is that a few times I have been able to get past the Works with Nest page when clicking "Accept", but then a mostly blank page is displayed with an 8 character string consitising of uppercase letters and numbers (Example: 1ADLRUDG). When I click "Done" in the app (upper right), I am sent back to the first Nest Manager page and the red "Click to enter Nest Credentials" button is still there. I am able to click the "Next" button in the app, but there are no Next Locations to select on the next screen. 

Any help would be appreciated. 

Thanks!

**Comment by tonesto7** [comment]
I'm sorry I can only do so much hand holding on this... 
You obviously missed a step even though you swear you didn't read the document clearly.   

All you need to do is copy and paste some text into a page and verify there are no extra spaces at the ends of the callback url and that it didn't add http:// in front of https:// url you pasted in.   

Then copy/paste into the smartapp's settings window in the IDE.  
If I could automate this part i really would.

**Comment by kahnman** [comment]
Thanks for your time @tonesto7. I went through all of the installation steps again, including creating the Nest product on their developer site and realized I missed entering https://graph.api.smartthings.com/oauth/callback as the OAuth Redirect URI there. Once I did that, I was able to continue with my configuration. Sorry for the trouble!

**Comment by mitchcat2000** [comment]
Ok, so I did just as kahnman did and did not have the https://graph.api.smartthings.com/oauth/callback in the nest Dev site. Thanks everyone for your help sorry to be a pain @tonesto7 but thanks for working through it with us. I will be sending in a donation. 

![untitled](https://user-images.githubusercontent.com/19941396/32675407-1d950872-c614-11e7-8280-e8e44ae52387.jpg)

**Comment by myrison** [comment]
@tonesto7 - thank you for replying and apologies for all of us asking questions about the same thing. I'm not sure what went wrong, but I deleted everything, started over, and it worked this time.

For others in the future wondering whether you have to use your "sharded" API URI (i.e. na04-useast2) for the oauth redirect URL, you do NOT.  This standard URI (https://graph.api.smartthings.com/oauth/callback) does work.

Sorry to have wasted your time... donation inbound from me as well.  The least I can do!

**Comment by japaho** [comment]
Also having this issue, oauth enabled, guide followed to the letter. Getting 'oops' when I try to sign in

**Comment by japaho** [comment]
This XML file does not appear to have any style information associated with it. The document tree is shown below.
<oauth>
<error_description>33499001-90ed-4885-899f-98e1bcc729a8</error_description>
<error>invalid_token</error>
</oauth>

**Comment by myrison** [comment]
@japaho - what I did the second time around when deleting everything and starting over was to copy and paste all device keys and redirect URLs first into an unformatted text document, and then copied each field back out of the text doc back into the associated field in the browser. 

My best guess (other than just dumb error on my part) was that some non-visible formatting or spacing was coming over during the copy/paste.  Maybe that'll help...

**Comment by tonesto7** [comment]
I want to add some JavaScript buttons to copy the stuff to the clipboard for you.  I will work on that when I have some time

**Comment by japaho** [comment]
Damn.. my bad.. @myrison you had it right, the copy paste from nest developer site adds a space to the end of the keys :-( all sorted now! Thanks for the tip off!

**Comment by myrison** [comment]
Awesome @japaho.  I never actually confirmed what the problem was when I redid everything, so it's good to know that was it.  Before I reinstalled everything, when I would go back into the web config page of SmartThings, I couldn't see the space there, but I figured it might be there regardless... Good to know that was actually what was happening so that @tonesto7 can add this note to the install instructions (i.e. copy into an unformatted text doc, then copy again and paste back into SmartThings).  No wonder so many people run into this issue during setup.

**Comment by biggles11** [comment]
I am getting We can no longer provide you with a built in Nest Client ID and Secret. Please Check the Wiki for Detailed instructions on creating your own Nest Dev ID and Secret.

I have created the Nest Dev ID and have the ID and secret, but still get this message. I assume I've missed a step butstill get this even after putting the clientID and secret in the settings of the app

**Comment by spencerthayer** [comment]
I too cannot click `accept` without receiving the `Oops! We've encountered an error. Please try again.` issue.

**Comment by ChrisSumner** [comment]
I'm also getting "Oops! We've encountered an error. Please try again." error message after recently enabling 2FA. Either 2FA isn't supported or run out of tokens. It would be great to find out which and if it's the latter, how long it will be until there will be a solution.

**Comment by joeyavino** [comment]
I'm getting the Oops! error too!  I've created my own Nest dev account, followed the instructions here: https://docs.google.com/document/d/1JoqNijFvLAHaQIMrqHCSBXgglUqdEnj3rBD9EiPP3bs/edit

And I ended in the Redirect URL that Nest provided, still no go.

Any thoughts?

Need help to get this working.

**Comment by EriKFL07** [comment]
I am getting the Oops error Too!  I have completed two factor authentication also.    
Nest Manager 5.2.9 - Have attempted to fix this for 2 days by uninstalling installing etc..

f567031b-e11e-422f-ba9d-cdf9a662af4e  9:21:12 PM: info ||| Nest Manager | AuthToken not found: Directing to Login Page
f567031b-e11e-422f-ba9d-cdf9a662af4e  9:20:00 PM: warn || Nest Manager | Missing: structData null

**Comment by imnotbob** [comment]
get 5.2.9 as of now (last 5 mins) and try again

**Comment by Electrician4Hire** [comment]
I'm getting the same error as erikfl07. I Updated too. I can get past the first login page and get to the nest works page. But when I click accept I get that oops error


47191d7e-9bd6-4e1f-a152-14c4fb989174  6:01:49 PM: debug Parse returned [name:temperature, value:63, unit:F, descriptionText:{{ device.displayName }} was {{ value }}°F, translatable:true]
47191d7e-9bd6-4e1f-a152-14c4fb989174  6:01:49 PM: debug description: temperature: 17.41
28b8a23c-4fe3-4c11-86df-9f6fea76b1bf  6:01:32 PM: debug pollChild(child)>> data for 28b8a23c-4fe3-4c11-86df-9f6fea76b1bf.511832992673 : [deviceAlive:true, deviceTemperatureUnit:F, heatMode:true, coolingSetpoint:72, minCoolingSetpoint:65, thermostatFanMode:auto, maxHeatingSetpoint:79, temperature:67.7, heatingSetpoint:68, minHeatingSetpoint:45, humidity:33, autoMode:true, thermostatMode:heat, coolMode:true, auxHeatMode:false, maxCoolingSetpoint:92]
28b8a23c-4fe3-4c11-86df-9f6fea76b1bf  6:01:32 PM: debug updated 1 stats: [28b8a23c-4fe3-4c11-86df-9f6fea76b1bf.511832992673:[data:[deviceAlive:true, deviceTemperatureUnit:F, heatMode:true, coolingSetpoint:72, minCoolingSetpoint:65, thermostatFanMode:auto, maxHeatingSetpoint:79, temperature:67.7, heatingSetpoint:68, minHeatingSetpoint:45, humidity:33, autoMode:true, thermostatMode:heat, coolMode:true, auxHeatMode:false, maxCoolingSetpoint:92]]]
28b8a23c-4fe3-4c11-86df-9f6fea76b1bf  6:01:32 PM: trace Storing thermostat data: [[lastModified:2018-01-09 12:00:18, settings:[serviceRemindMe:false, stage1CoolingDifferentialTemp:5, dehumidifierLevel:60, remindMeDate:2014-04-28, drAccept:always, disableHeatPumpAlerts:false, coldTempAlertEnabled:true, compressorProtectionMinTemp:350, hvacMode:heat, heatStages:1, dehumidifyWithAC:false, disablePreHeating:false, auxMaxOutdoorTemp:700, maxSetBack:100, humidityHighAlert:-1, backlightOnIntensity:10, humidifierMode:off, humidity:36, dehumidifyWhenHeating:false, compressorProtectionMinTime:300, soundAlertVolume:0, heatMaxTemp:1200, ventilatorMinOnTime:20, installerCodeRequired:false, groupName:, hasHeatPump:false, holdAction:askMe, backlightOffDuringSleep:false, hotTempAlertEnabled:true, humidityAlertNotifyTechnician:false, backlightSleepIntensity:4, randomStartDelayHeat:0, useCelsius:false, auxRuntimeAlertNotifyTechnician:false, stage1HeatingDissipationTime:31, smartCirculation:false, tempAlertNotifyTechnician:false, coolMaxTemp:1200, humidityLowAlert:-1, humidityAlertNotify:false, autoAway:true, randomStartDelayCool:0, vent:off, heatPumpReversalOnCool:true, fanControlRequired:true, soundTickVolume:0, isRentalProperty:false, hasForcedAir:true, useZoneController:false, ventilatorDehumidify:true, fanMinOnTime:0, dehumidifyOvercoolOffset:0, heatMinTemp:450, quickSaveSetBack:40, stage1HeatingDifferentialTemp:5, hasBoiler:false, coolRangeHigh:920, backlightOffTime:60, auxOutdoorTempAlertNotifyTechnician:false, hasUVFilter:true, eiLocation:, quickSaveSetForward:40, hasErv:false, heatRangeLow:450, disableAlertsOnIdt:false, coolMinTemp:-100, ventilatorMinOnTimeHome:20, coolingLockout:false, auxOutdoorTempAlert:500, stage1CoolingDissipationTime:31, enableElectricityBillAlert:false, followMeComfort:true, lastServiceDate:2014-04-28, condensationAvoid:false, groupRef:, auxOutdoorTempAlertNotify:false, electricityBillStartMonth:1, ventilatorOffDateTime:, hasElectric:false, autoHeatCoolFeatureEnabled:true, monthsBetweenService:6, locale:en, ventilatorFreeCooling:true, heatRangeHigh:790, ...[TRUNCATED]
28b8a23c-4fe3-4c11-86df-9f6fea76b1bf  6:01:32 PM: debug updated 1 thermostats: [28b8a23c-4fe3-4c11-86df-9f6fea76b1bf.511832992673:[data:[deviceAlive:true, deviceTemperatureUnit:F, heatMode:true, coolingSetpoint:72, minCoolingSetpoint:65, thermostatFanMode:auto, maxHeatingSetpoint:79, temperature:67.7, heatingSetpoint:68, minHeatingSetpoint:45, humidity:33, autoMode:true, thermostatMode:heat, coolMode:true, auxHeatMode:false, maxCoolingSetpoint:92]]]
28b8a23c-4fe3-4c11-86df-9f6fea76b1bf  6:01:32 PM: debug polling children: 511832992673
28b8a23c-4fe3-4c11-86df-9f6fea76b1bf  6:01:32 PM: debug getChildDevices(false), children=2
47191d7e-9bd6-4e1f-a152-14c4fb989174  5:56:49 PM: debug Parse returned [name:temperature, value:63, unit:F, descriptionText:{{ device.displayName }} was {{ value }}°F, translatable:true]
47191d7e-9bd6-4e1f-a152-14c4fb989174  5:56:49 PM: debug description: temperature: 17.46
29160de6-a2e0-416e-84ac-07ef65a8d72b  5:56:08 PM: warn || Nest Manager | Missing: structData null
29160de6-a2e0-416e-84ac-07ef65a8d72b  5:56:05 PM: info ||| Nest Manager | AuthToken not found: Directing to Login Page
 5:55:59 PM: info Waiting on events...
 5:55:59 PM: info For past logs for individual things go to the My Devices section, find the device and click on the Events link on the device information page.
 5:55:59 PM: info This console provides live logging of your SmartThings.

**Comment by imnotbob** [comment]
ask WB70 in the forums to have a look, he has been good at initial scrub of issues.

**Comment by bancroftsinha** [comment]
Opps! something went wrong error

**Comment by Jamessmart1** [comment]
Opps! something went wrong error. please try again. running nest manager ver. 5.3.2

**Comment by Jamessmart1** [comment]
using nest manager 5.3.2 updated:  01-13-2018. I can log online no problems, and the app no problems but when i log into the smartthing app choose my app, nest manager,and then log onto nest credentials I get this error. Opps! something went wrong error

**Comment by imnotbob** [comment]
It is important that you create your own nest ID and secret.  

It is important that when following the instructions, you are careful in filling in field to ensure no extra characters / spaces were put in.

See above comment on folks correcting their issues and things working

**Comment by Jamessmart1** [comment]
thanks

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=icon>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=link>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by sathyanatarajan76** [comment]
I tried updating my smart app and device handler and stuck with the oops error when trying to login. Pls help.

**Comment by scuk07** [comment]
I'm getting the same. Followed the installation instructions and initially got a Client ID / Client Secret error. I went into Edit on the Nest Manager SmartApp, copied the ID and secret from the OAuth section to the clientID and clientSecret fields in Settings, which got me a little further, but now when I fire up the SmartApp in SmartThings, click to Login to Nest and enter my details I also get the 'Opps! We've encountered an error. Please try again.' error.

I've confirmed my Nest login, reset my Nest password and tried again, and even stripped out the SmartApp and Device Handlers, reset my SmartThings hub and gone through the entire NST Manager instructions again. No joy, same error.

Does anyone have any thoughts as to what I'm missing?

**Comment by dwightmb** [comment]
Here is what worked for me.  I had to add an additional OAuth Redirect to my Nest Developers account.  I added the following https://graph.api.smartthings.com/oauth/callback  closed and relaunched the smartthings app and then it logged in no problems.  This resolved my "ops something went wrong error".  Hope it helps.

**Comment by scooter3** [comment]
@dwightmb Been looking all over the place for a solution. That worked for me. Thanks.

**Comment by chinadollmk** [comment]
Having the same issue after entering login credentials for Nest “Oops! We’ve encountered an error. Please try again.” I have the Nest Client ID and Client Secret creates and entered into settings. OAuth is enabled, code is updated and published. Can’t uninstall the SmartApp code from my IDE because it’s been installed with one or more users. Any ideas on how to get this working?

**Comment by imnotbob** [comment]
If all the instructions are followed carefully, things work is our experience.   If you are sure all details are correct,  you can look a live logging during "a try" to see if the logs say anything.

Otherwise you may want to contact ST support.   Most of the time careful following of the instructions (and careful cut/paste) resolves things.   Tony and I have not made changes recently that might affect things.

**Comment by imnotbob** [comment]
restarting the ST mobile app (or your phone) has helped some folks also.

We are sorry for the trouble, ST has many moving parts, and sometimes they don't move so well due to [temporary] platform issues.

**Comment by chinadollmk** [comment]
I uninstalled Nest Manager from IDE. Reinstalled, then try to set location. Looked at the logs and looks like there’s no AuthToken found which is throwing the error. Any ideas on how to fix?

**Comment by imnotbob** [comment]
We likely would need to see the exact error message

At first glance, it appears the nest developer site side setup and/or the clientId and clientSecret settings in the IDE are not correct.  You should check that the instructions on the nest developer side (and spellings) are all exactly correct, and that the clientId and clientSecret are correctly set in the ST IDE appSettings.

I assume it takes you thru the Nest site to login then you get the error?

**Comment by chinadollmk** [comment]
double-checked client ID and client Secret, client ID was incorrect, so corrected and now it works great! thanks!

**Comment by donthategg** [comment]
Hey All- I'm trying to set all this up, but upon signing up for Nest Developers site, received a notification saying Nest has paused and is restructuring its developr program, and no longer accepting new developers.  Is there a workaround for a token?

**Comment by cgross86** [comment]
donthategg,

Thought Im stuck there as well however I punshed in my info, followed the Nest developer instructions, got the Client and SecretID, updated the settings in Smartthings and it now works.

**Comment by scoob5684** [comment]
I think I'm having a token problem

**Comment by timeisanillusion** [comment]
Same issue.  I've removed and reinstalled without luck
Logs show
9:47:56 AM: warn || Nest Manager (v5.6.4) | Missing: structData null
9:47:52 AM: info ||| Nest Manager (v5.6.4) | AuthToken not found: Directing to Login Page
9:47:52 AM: info ||| Nest Manager (v5.6.4) | appConfig.json File HAS Changed

**Comment by Ibanez1998** [comment]
I am having the issue where I can't allow works with Nest.  I think another token upgrade needs completing.  Please advise.

**Comment by myrison** [comment]
Google disabled this service after acquiring Nest. It is no longer possible unfortunately.

**Comment by Ibanez1998** [comment]
> Google disabled this service after acquiring Nest. It is no longer possible unfortunately.

Oh NO!!!  :(
